File: src/db/logs_repo.py
"""
Repository for logs table operations.
"""

import sqlite3
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Optional

ISO_FMT = "%Y-%m-%dT%H:%M:%S"


@dataclass
class LogRow:
    """Log record from DB."""

    id: int
    level: str
    message: str
    context: Optional[str]
    created_at: str


def insert(
    conn: sqlite3.Connection,
    level: str,
    message: str,
    context: Optional[str] = None,
) -> int:
    """Insert log. Returns new id."""
    now = datetime.now(timezone.utc).strftime(ISO_FMT)
    cursor = conn.execute(
        "INSERT INTO logs (level, message, context, created_at) VALUES (?, ?, ?, ?)",
        (level, message, context, now),
    )
    conn.commit()
    return cursor.lastrowid


def list_all(
    conn: sqlite3.Connection,
    level: Optional[str] = None,
    limit: int = 200,
) -> list[LogRow]:
    """List logs, optionally filtered by level."""
    if level:
        rows = conn.execute(
            "SELECT id, level, message, context, created_at FROM logs WHERE level = ? ORDER BY id DESC LIMIT ?",
            (level, limit),
        ).fetchall()
    else:
        rows = conn.execute(
            "SELECT id, level, message, context, created_at FROM logs ORDER BY id DESC LIMIT ?",
            (limit,),
        ).fetchall()
    return [
        LogRow(id=r[0], level=r[1], message=r[2], context=r[3], created_at=r[4])
        for r in rows
    ]

File: src/db/logs_repo.cpp
#include "document_ingestion/logs_repo.hpp"
#include <sqlite3.h>

namespace document_ingestion {

void log_insert(sqlite3* conn, const std::string& level, const std::string& message,
                const std::string* context) {
  sqlite3_stmt* stmt = nullptr;
  const char* sql = "INSERT INTO logs (level, message, context) VALUES (?, ?, ?)";
  if (sqlite3_prepare_v2(conn, sql, -1, &stmt, nullptr) != SQLITE_OK) return;
  sqlite3_bind_text(stmt, 1, level.c_str(), -1, SQLITE_TRANSIENT);
  sqlite3_bind_text(stmt, 2, message.c_str(), -1, SQLITE_TRANSIENT);
  if (context)
    sqlite3_bind_text(stmt, 3, context->c_str(), -1, SQLITE_TRANSIENT);
  else
    sqlite3_bind_null(stmt, 3);
  sqlite3_step(stmt);
  sqlite3_finalize(stmt);
}

}  // namespace document_ingestion
